runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerStepMap.java

```diff
@@ -32,55 +30,53 @@
 import org.apache.beam.sdk.metrics.MetricKey;
 import org.apache.beam.sdk.metrics.MetricResult;
 import org.apache.beam.sdk.metrics.MetricResults;
+import org.apache.beam.sdk.metrics.labels.MetricLabels;
 
 /**
  * Metrics containers by step.
  *
  * <p>This class is not thread-safe.
  */
 public class MetricsContainerStepMap implements Serializable {
-
-  private Map<String, MetricsContainerImpl> metricsContainers;
-  private MetricsContainerImpl unboundContainer = new MetricsContainerImpl(null);
+  private Map<MetricLabels, MetricsContainerImpl> metricsContainers;
 
   public MetricsContainerStepMap() {
     this.metricsContainers = new ConcurrentHashMap<>();
   }
 
-  /* Returns the container that is not bound to any step name. */
-  public MetricsContainerImpl getUnboundContainer() {
-    return this.unboundContainer;
+  public MetricsContainerImpl ptransformContainer(String ptransform) {
+    return getContainer(MetricLabels.ptransform(ptransform));
+  }
+
+  public MetricsContainerImpl pcollectionContainer(String pcollection) {
+    return getContainer(MetricLabels.pcollection(pcollection));
   }
 
   /** Returns the container for the given step name. */
-  public MetricsContainerImpl getContainer(String stepName) {
-    if (stepName == null) {
-      // TODO(BEAM-6538): Disallow this in the future, some tests rely on an empty step name today.
-      return getUnboundContainer();
-    }
-    if (!metricsContainers.containsKey(stepName)) {
-      metricsContainers.put(stepName, new MetricsContainerImpl(stepName));
+  public MetricsContainerImpl getContainer(MetricLabels labels) {
+    if (!metricsContainers.containsKey(labels)) {
```

Comment:
It doesnt seem like `MetricLabels` implements a `hashCode` or an `equals` function. I may be missing something, but perhaps we need a test that checks `getContainer(MetricLabels.ptransform("A")).equals(MetricLabels.ptransform("A"))` ?